File: 1-easy/problems/0234-palindrome-linked-list/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def isPalindrome(head: ListNode) -> bool:
    if not head or not head.next:
        return True

    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    prev = None
    while slow:
        next_temp = slow.next
        slow.next = prev
        prev = slow
        slow = next_temp

    while prev:
        if prev.val != head.val:
            return False
        prev = prev.next
        head = head.next

    return True

File: 1-easy/problems/0234-palindrome-linked-list/solution.ts
function isPalindrome(head: ListNode | null): boolean {
    if (!head || !head.next) return true;

    let slow = head;
    let fast = head;
    while (fast && fast.next) {
        slow = slow.next!;
        fast = fast.next.next;
    }

    let prev: ListNode | null = null;
    while (slow) {
        const nextTemp = slow.next;
        slow.next = prev;
        prev = slow;
        slow = nextTemp;
    }

    while (prev) {
        if (prev.val !== head!.val) return false;
        prev = prev.next;
        head = head!.next;
    }

    return true;
}
